Make update request handler configurable

ojlyytinen opened the pull request:
Adds support to use other than the default '/update' and '/select' request handlers in Solr.

ojlyytinen commented:
Not sure what's going wrong with Travis. It's only Ruby 2.2.0 and Ruby 1.9.3 that are failing and they seem to fail at ```bundle install```. The error message is ```NoMethodError: undefined method `spec' for nil:NilClass```

jcoyne commented:
I believe that's a bug in bundler version 1.8. Probaby fixed by https://github.com/bundler/bundler/commit/5749adedea810a94dc5c156f0f36247c572ea368

jcoyne commented on lib/rsolr/client.rb:
```diff
@@ -63,6 +65,11 @@ def paginate page, per_page, path, opts = nil
   end
   
+  # Sends a query to /select or the custom configured select handler.
+  def select opts={}
```
I don't see the need for this. Why not just call `get '/select'` or whatever handler you want to use here. the `/select` handler is not something you have to use, so RSolr has no need to know about this.

ojlyytinen replied:
The idea is that whatever is using RSolr doesn't need to know which handler to use. The ActiveFedora PR has [this](https://github.com/projecthydra/active_fedora/pull/960/files#diff-737ca785dbb0fa8180646721bf05d662R108) bit. So it's doing a select and using whatever is configured to be the select handler in RSolr. If we didn't have a select method here then ActiveFedora would need to dig into Solr configuration to see which handler needs to be used.

I've taken the approach of having all Solr configuration tied to RSolr. It seemed that the RSolr::Client object is the better place to carry the information about handlers since it's abstracting the connection to Solr. Certainly it would be possible to instead have ActiveFedora know what handler needs to be used and then have it specify that when doing a select or an update.

jcoyne replied:
This Gem isn't used just by ActiveFedora, so we don't want to add features that are just applicable to ActiveFedora in here.

ojlyytinen replied:
No of course not, that's just an example of what the purpose of the select method would be instead of using ```get '/select'```. Anything that uses RSolr can have the Solr settings in RSolr instead of having the logic to select handlers outside of it.

In any case, if you prefer keeping the logic of selecting which handler to use outside RSolr I can certainly refactor these two pull requests that way.

cbeer replied:
I think that makes sense -- `select` is a convention, and not a configuration, while `update` is configuration and incorrectly hard-coded.

ojlyytinen replied:
Right, I think I see your point. I'll remove the select method and configuration from here and move that stuff in ActiveFedora. 

Would you rather have the update_path here passed to initialize, like it is now, or should I change that to be in the opts parameter of update instead?

cbeer replied:
I think supporting both an instance-level and method-level configuration would be idea

jcoyne commented on lib/rsolr/client.rb:
```diff
@@ -79,7 +86,7 @@ def paginate page, per_page, path, opts = nil
     opts[:headers] ||= {}
     opts[:headers]['Content-Type'] ||= 'text/xml'
-    post 'update', opts
+    post update_path, opts
```
This part seems like a good idea.

ojlyytinen commented:
I've changed this PR as per earlier discussion and rebased. This PR now only makes the update handler configurable and nothing else.

Travis is still having issues with ```bundle install``` and in addition one test is also failing. This test seems to fail on master branch as well when I run the specs locally.

cbeer commented on lib/rsolr/client.rb:
```diff
@@ -79,7 +80,7 @@ def paginate page, per_page, path, opts = nil
     opts[:headers] ||= {}
     opts[:headers]['Content-Type'] ||= 'text/xml'
-    post 'update', opts
+    post opts.fetch(:update_path, update_path), opts
```
What do you think about calling this option just `path`? I see `update` is also called below by a couple other methods, and the connection between `update_path` and, say, `commit` may not be obvious.

ojlyytinen commented:
Yes, ```:path``` as the opts key for ```update``` makes more sense. The same key is used in ```build_request``` but I don't think that matters as it'll just overwrite it with the same value.

cbeer commented:
:+1: Thanks.
